Add draft tutorial for jni programming

johnwalker opened the pull request:
Hi there, this is a followup draft from #222 . I'd like to work with you on the content here. A few others contributed to these materials as well. :)

## Overview

Add a tutorial that introduces JNI programming with jni-rs. Build with mdbook to get a preview. One of the goals of the book is to have the reader do stuff and think about JNI/FFI in exercises. At the moment, it introduces JNI with division, debugging, and a native Counter implementation. You will see TODOs as you reach the end of the tutorial, especially on Cleaners.

### Definition of Done

- [x] There are no TODOs left in the code
- [ ] Everyone is happy enough with writing style and content
- [x] The material is tested on Windows
- [x] The material is tested on Linux
- [x] The material is tested on OSX
- [ ] The authors tag is filled in
- [ ] Change is covered by automated [tests](https://github.com/jni-rs/jni-rs/blob/master/CONTRIBUTING.md#tests)
- [ ] The [coding guidelines](https://github.com/jni-rs/jni-rs/blob/master/CONTRIBUTING.md#the-code-style) are followed
- [x] Public API has documentation
- [x] This change is not breaking **or** mentioned in the Changelog
- [ ] The [continuous integration build](https://www.travis-ci.org/jni-rs/jni-rs) passes

dmitry-timofeev commented on book/src/linking_and_dividing.md:
````diff
@@ -0,0 +1,337 @@
+to be that way, it's just a convention that we follow here.
+
+```java
+package com.github.jni_rs.jnibook;
````
Is it possible to include code from files?

dmitry-timofeev commented on book/src/linking_and_dividing.md:
```diff
@@ -0,0 +1,337 @@
+       try {
+           System.loadLibrary("jnibookrs");
+       } catch (Throwable t) {
+           error = t;
```
(here and on the full file)

Why don't we throw an Exception, with, potentially, a clearer error message,
on the first use? It seems to be simpler, fails immediately, and there is no need to repeatedly checkAvailability which never changes after the static initializer has executed: the library was either loaded or not.

For better error reporting (non tutorial purposes), it might make sense to extract that code, common across several classes with native methods, in a separate class. E.g., that's what Exonum has (it has to be clear on what's wrong as a framework): https://github.com/exonum/exonum-java-binding/blob/ed61d3454469530575c264074ed9ed06c0300039/exonum-java-binding/core/src/main/java/com/exonum/binding/core/util/LibraryLoader.java#L115

dmitry-timofeev replied:
Mostly, simplicity, and also the fact that you get an error as soon as it occurs. Or are you saying that the benefit of this approach is that any subsequent usages (e.g., in tests) will include the original message (without this, you will have to find the _first_ error message).

dmitry-timofeev commented on book/src/linking_and_dividing.md:
```diff
@@ -0,0 +1,337 @@
+`.` is mapped to `_`. However, since `_` could appear in a function name, it
+needs a different representation. So `_` gets replaced by `_1` before replacing
+`.` with `_`. And finally, the function name has to be prefixed with `Java_`, to
+avoid conflicts with other languages that support similar naming schemes.
```
Wouldn't it be better to suggest that users run `javac -h <path-to-headers>`
and copy the function declarations from the generated header file(s)?

E.g.,
```
javac -h . NativeAPI.java
cat NativeAPI.h

# Or with Gradle
./gradlew build
cat target/<our-path>/NativeAPI.h
```

It doesn't have to be configured separately from usual javac invocation,
see for example, https://github.com/exonum/exonum-java-binding/blob/master/exonum-java-binding/core/pom.xml#L184). It can be put into our Gradle template, so that in `target/native-headers`
there are up-to-date headers with correct function names.

javac reliably produces correct method names, that a user can copy and paste.
Even though the transformations below are described really well, and the examples
illustrate them well, if someone tries to follow them IRL, it may be tedious:
  - Copy/type the package name
  - Separately copy/type method name
  - Mangle as specified.

I think the rules distract from the more interesting parts; and may be harmful:
it is possible to mis-understand them, or make a typo or an error during a manual
conversion, or miss "Java_", and end up debugging "LinkageError: symbol not found".

Surely, it would be lovely to not have to do that at all: https://github.com/jni-rs/jni-rs/issues/81

johnwalker replied:
Decided to just use javac for now

dmitry-timofeev commented on book/src/linking_and_dividing.md:
```diff
@@ -0,0 +1,337 @@
+
+Now the native function is named correctly for JNI.
+
+## Parameters and Return Types
```
Would it be clearer if first a 1st argument (`JNIEnv`) is desribed; 
then — the 2nd argument:
- for native static methods (a reference to the defining class);
- for native instance methods (a reference to `this`)?

johnwalker replied:
Yes it would. The writing there is kind of odd, in that it talks about argument 2, then 1, then 2. Will fix.

johnwalker replied:
Updated this part

dmitry-timofeev commented on book/src/linking_and_dividing.md:
```diff
@@ -0,0 +1,337 @@
+not for shared libraries.
+
+Therefore, it's necessary to explicitly set the ABI using `pub extern "system"`.
+For further information, see https://doc.rust-lang.org/std/keyword.extern.html
```
My mdbook renders them as plain text :confused:

johnwalker replied:
Ah, got it. I think I fixed similar problems through the rest of the tutorial.

dmitry-timofeev commented on book/src/linking_and_dividing.md:
````diff
@@ -0,0 +1,337 @@
+}
+```
+
+Finally, set the ABI using `pub extern “system”`. 
````
Fancy quotes in code

johnwalker replied:
Fixed all instances

dmitry-timofeev commented on book/src/linking_and_dividing.md:
```diff
@@ -0,0 +1,337 @@
+
+`JNIEnv` instances share the lifetime of the calling thread, which means you
+should avoid holding onto `JNIEnv` instances. Once the calling thread is
+deallocated, it the `JNIEnv` instance will no longer be valid. [JNI00]
```
probably, redundant "it"

johnwalker replied:
Fixed

dmitry-timofeev reviewed (commented):
I've reviewed only the first two chapters, but brilliant job! Some comments below.

johnwalker commented on book/src/introduction.md:
```diff
@@ -0,0 +1,169 @@
+# Introduction
+This book is about building shared libraries in Rust that work with Java via JNI
+(Java Native Interface). JNI is a Java API for using shared libraries from Java.
```
Thanks, I like that description much more. :)

johnwalker replied:
I borrowed the wording you suggested and rewrote the intro to talk about applications and libraries

johnwalker commented on book/src/introduction.md:
```diff
@@ -0,0 +1,169 @@
+layer, so that all benefits you get from Rust in normal application or library
+development apply to the core. Within the JNI layer itself, similar benefits
+still apply with the caveat that JNI API misuse is still highly possible, and
+unsafe Rust is often necessary. Expect the JNI code for Rust to look similar to
```
I decided to remove discussion of this, since often depends entirely on what you're trying to do and may confuse people.

johnwalker commented on book/src/introduction.md:
```diff
@@ -0,0 +1,169 @@
+For Java, you need experience with defining classes. For Rust, you should have
+experience with `Box`, and defining `struct`s and `function`s.
+
+Your system needs Java 11+, recent versions of Cargo and Rust, and one Java
```
I rewrote this part to explicitly mention that jni-rs is compatible with older versions, and that it's strictly a tutorial dependency.

johnwalker commented:
@dmitry-timofeev thoughts so far? I'm thinking of fleshing out the memory section and maybe adding an example for using hyper over jni.

dmitry-timofeev commented:
@johnwalker , sorry for the massive delay! I'll try to review the updates and the next chapters in the coming weeks. I've recently switched jobs, and no longer work for a company-active jni-rs user, so can only contribute in my personal capacity. 

If your project or company starts using jni-rs extensively, please consider subscribing for updates and getting involved in reviews/design discussions. This project was originally created by @jrobsonchase for Prevoty, and then was led by Exonum Java Binding team, with many contributions from engineers from other projects. Having feedback from direct usage helps a lot.

dmitry-timofeev commented on book/src/introduction.md:
```diff
@@ -0,0 +1,169 @@
+to resource leaks or JVM crashes. There are also performance implications of
+using JNI in certain ways. For example, Java Strings and Buffers sometimes have
+to be copied when they're used in the shared library. Due to the drawbacks of
+JNI, it's usually more appropriate to build using pure Java.
```
I think JNR allows to use native libraries without any glue code (you define only a Java interface and annotate it). There is also JNA. 

The difference between them, if I remember correctly, is that JNR generates the glue code in object form (so it may be as efficient as hand-written JNI); whilst JNA uses a native library with one method that performs dynamic invocation by symbol name.

dmitry-timofeev commented on book/src/cleaning.md:
```diff
@@ -0,0 +1,62 @@
+
+Since Java doesn’t automatically free the memory on Rust’s heap when the
+`Counter` garbage collected, there could be a memory leak. Many native resources
+need similar treatment, and so we are going to automate it using the [Cleaner
```
Even if we don't cover the approach with explicit memory management, I think it shall be mentioned with a todo to expand on this topic, with some considerations on choosing the right approach.

Cleaner, as any GC-based approach, is not universally applicable:
- Cleaning actions are performed in a separate thread, hence the Java class has to be thread-safe
- If the app creates Java proxies of dependent native objects, they likely have to be destroyed in the reverse order of their creation — something that is possible with PhantomReferenceQueues, but will add quite some complexity
- Any finalization-based approach is not suitable for high-volume de-allocation rate. I no longer have the benchmarks, as they are in private Exonum repository, but about three years ago we also considered which approach to take o simplify the memory management, and prototyped:
    - Explicit (it may be eventually hidden in a single place if all native peers are created in a certain bounded context — that's how EJB works)
    - GC-based approaches:
      - `finalize`
      - `PhantomReferenceQueue`, no explicit clean action
      - `PhantomReferenceQueue`, with explicit clean action

Any GC-based approaches were at least an order of magnitude slower. The GC-based approaches compared in the following way, from fastest to slowest:
1. `PhantomReferenceQueue`, with explicit clean action
2. `finalize`
3. `PhantomReferenceQueue`, no explicit clean action
      
See also [my answer](https://stackoverflow.com/a/44799935) from the time we researched this topic in Exonum.

dmitry-timofeev replied:
Also, we can consider finishing this chapter in subsequent PRs, so that it doesn't grow too large.

johnwalker replied:
Yeah, we'll need to expand on this over time. I'll edit it to state that the chapter isn't a complete guide to memory management of native resources, and describe the upsides/downsides of the approach, then mention the alternatives you've described. Then, we can expand on those over time - potentially in chapters that revisit the same topic, or by additions to this particular chapter.

dmitry-timofeev commented on book/src/counter.md:
```diff
@@ -0,0 +1,331 @@
+    public int increment() {
+    }
+
+    public void close() {
```
Shall we clarify here or in the previous paragraph what this method is supposed to do (destroy the native peer)?

johnwalker replied:
Yes, let's do that.

dmitry-timofeev commented on book/src/counter.md:
```diff
@@ -0,0 +1,331 @@
+    public int increment() {
+    }
+
+    public void close() {
```
I'd consider implementing AutoCloseable so that this class can be used in try-with-resources statement.

johnwalker replied:
Yeah, it's just a question of when best to do it. At the moment, Counter gets AutoCloseable in the cleaning chapter. I think we could probably introduce that here, though.

dmitry-timofeev commented on book/src/counter.md:
````diff
@@ -0,0 +1,331 @@
+```java
+class Counter {
+    // The pointer to the counter in the native (Rust's) heap
+    private long ptr;
````
final?

johnwalker replied:
fixed

dmitry-timofeev commented on book/src/counter.md:
```diff
@@ -0,0 +1,331 @@
+2. `counter_1new`, `counter_1increment`, and `counter_1get` should not deallocate the counter.
+3. `counter_1destroy` should deallocate the counter.
+
+We'll need to use [`Box`](https://doc.rust-lang.org/std/boxed/index.html) to
```
I wish we already had https://github.com/jni-rs/jni-rs/issues/84 moved from EJB into a satellite library to not have to discuss this :)

johnwalker replied:
Yeah that does look helpful. Fortunately, it's not a one-way door. Once #84 is resolved, we can decide how to introduce it here.

dmitry-timofeev commented on book/src/counter.md:
```diff
@@ -0,0 +1,331 @@
+
+We'll need to use [`Box`](https://doc.rust-lang.org/std/boxed/index.html) to
+meet these goals. `Box` is used for storing objects on the heap. Boxed objects
+can't outlive their boxes, except through through `Box::into_raw` and
```
```suggestion
can't outlive their boxes, except through `Box::into_raw` and
```

dmitry-timofeev commented on book/src/counter.md:
```diff
@@ -0,0 +1,331 @@
+boundaries, since they enable the other language to play a role in memory
+management. By virtue of letting multiple languages play together with memory,
+it's easier to make memory errors at this boundary than elsewhere - either in
+pure Java, or in Pure Rust.
```
```suggestion
pure Java, or in pure Rust.
```

dmitry-timofeev commented on book/src/counter.md:
```diff
@@ -0,0 +1,331 @@
+## Implement Increment and Get
+
+To implement increment, we'll need to convert `ptr` back into a `Box<Counter>`,
+and then call increment, drop the box through `into_raw` to about deallocating
```
avoid deallocating?

dmitry-timofeev commented on book/src/counter.md:
```diff
@@ -0,0 +1,331 @@
+## Exercises
+
+1. What happens if destroy is never called, or gets called multiple times?
+2. Does the memory created by Rust count towards heap size limits set through
```
The ultimate answer from async-profiler author, Andrei Pangin: https://www.youtube.com/watch?v=c755fFv1Rnk :-)

dmitry-timofeev commented on book/src/debugging.md:
```diff
@@ -0,0 +1,63 @@
+
+You'll need to find the PID of your test. If you're using Java 11, then you can
+fetch it using the `ProcessHandle` API, as shown below. Otherwise, you'll need
+to find the PID through other means, such as through `ps -x` on Unix-like
```
There is [`jps`](https://docs.oracle.com/javase/9/tools/jps.htm#JSWOR733) coming with any JDK, possibly, it is also available on Windows.

The docs say though "experimental and unsupported" :)

dmitry-timofeev commented on book/src/error_handling.md:
```diff
@@ -0,0 +1,57 @@
+2. The function
+   [`catch_unwind`](https://doc.rust-lang.org/std/panic/fn.catch_unwind.html)
+   will allow you to execute code and recover from a call to `panic!`. This is
+   discouraged since it won’t catch all panics (in some cases, a panic may just
```
Is it _generally_ discouraged, or is it also discouraged in this use-case of catching panics at the FFI boundary? My understanding is that it is fine to use at FFI boundary to prevent UB. In Exonum we did catch panics, and then converted them into Exceptions: https://github.com/exonum/exonum-java-binding/blob/master/exonum-java-binding/core/rust/src/storage/list_index.rs#L54
    
If it's fine (and even, recommended) to use here, I'd clarify that.

dmitry-timofeev commented on book/src/java_objects_in_rust.md:
```diff
@@ -0,0 +1,75 @@
+    3. Use the
+       [`AutoLocal`](https://docs.rs/jni/0.15.0/jni/struct.JNIEnv.html#method.auto_local)
+       helper implemented within the JNI crate, so you don’t have to manually
+       track the usage of that reference at all.
```
I think a rule of thumb is that if you create more than a couple of AutoLocals,
       with_local_frame will be more efficient (I think we have benchmarks for that in `//benches/`),
       therefore it's not like AutoLocals are always better. If docs are unclear on that, it might be
       better to update them instead of repeating them here.

johnwalker replied:
The `AutoLocal` docs (https://docs.rs/jni/0.18.0/jni/objects/struct.AutoLocal.html) don't currently mention `with_local_frame`. I'd also expect `with_local_frame` to be more efficient, and think that's worth mentioning in the rustdocs.

Is there any way I can add this as a task to the project board? I don't think I have permissions to manipulate it.

dmitry-timofeev replied:
> Is there any way I can add this as a task to the project board? I don't think I have permissions to manipulate it.

I've sent a collaborator invitation earlier, but see it is still pending. Likely, it must be somewhere in the mail. I think when you accept you must be able to modify the board.

johnwalker replied:
Aha, found it. Thank you :)

dmitry-timofeev commented on book/src/java_objects_in_rust.md:
```diff
@@ -0,0 +1,75 @@
+*only as long as* those references are shorter-lived than the `AutoLocal`
+itself.
+
+This is a contrived example (we would probably convert to `long` on the Java
```
My concern is the same: we show an example that doesn't seem to make sense
(Java loop over NativeCounters will be faster, because in that case no "Rust -> Java"
calls will happen; passing an array of longs may or may not be faster than a Java loop, but certainly will be faster than the example).

Maybe passing an array of Points to print/render them? Again, Points as Java objects might not be the most efficient solution in such a case :)

johnwalker replied:
Part of the motivation is to familiarize the reader with arrays of objects, and I agree with you that the solution doesn't make much sense for the problem considering all the extra Rust->Java calls. On the other hand, I like that it introduces field and array access.

We could probably replace NativeCounter with a different class (I'm open to it) or perform a different operation. I'm partly tempted by your suggestion to pass an array of Points or longs, since we could maybe move it to construction time (potentially not the critical path) and talk a little bit better about the cost of the JNI upcalls.

dmitry-timofeev replied:
> Again, Points as Java objects might not be the most efficient solution in such a case :)

Possibly, some objects with behaviour (not plain data holders, as with Points readers could see more efficient approaches), e.g., Java passing to native an array of `Spaceship`s, and native `Spaceship#launch`ing them.

johnwalker replied:
For an outline of this chapter, how do you feel about:

- Problem description: Describe as an exercise for updating many X (possibly Counters)
- Solution 1: Java loop. Describe as simplest solution.
- Solution 2: Rust loop with ptrs. Describe as more complex, but potentially warranted if performance tests demonstrated an improvement.
- Solution 3: Rust loop with Java objects. Describe as most complex, with the highest number of upcalls. Mention that in certain cases, JNI performance considerations may be negligible (As in , `Spaceship#launch`).
- Analysis: Summarize tradeoffs for each solution. (Number of upcalls/downcalls in each, complexity, additional code, flexibility for future changes). Conclude with an author's preference for 1 based off simplicity. Note that which of Solution 1 and 2 is faster depends on the JNI implementation, and list some considerations (for example, array copying).

As a result, the reader should get exposed to a few different patterns for managing arrays across Rust and Java, and they would be more inclined to reach for Solution 1, or try Solution 2 after testing performance themselves. However, they'd also be able to get data out of a Java object and LongArrays if they needed to.

dmitry-timofeev replied:
Depends on what this chapter attempts to cover. Is the focus on using Java objects in Rust, or accessing Java arrays?

For instance, if the focus is on passing ~~Java objects~~ data, I think we can make an example with a single argument, and consider two "types" of objects the native might need:
- Plain data carriers. In this case, accessing Java objects may or may not be the best idea (some of the considerations are the speed of development and the runtime efficiency). There are at least three viable approaches:
  - Pass Java objects as-is. Native -> Java method invocations are rather expensive; don't have data on field accesses. However, sometimes it might be simpler than other approaches in some cases.
  - Unroll a Java object into a collection of primitive fields (e.g., instead of passing a single `3dPoint`, pass `x, y, z`).
  - Pass the Java object in a serialized form as a byte array, e.g., a protocol buffer. Depending on the number and types of fields and serde performance, might or might not be better.
- Objects with behaviour that Rust _needs_ to call (e.g., Java object implements a Rust trait). In this case, no much choice here: you have to pass a `Spaceship` if the `MissionControlCenter` is in Rust.

If we limit tje discussion to objects (and do not discuss passing data), there will be only one solution; if extend to data: at least three.

I'd consider keeping arrays as a separate sub-section, if we discuss passing data.

johnwalker replied:
Based off this feedback, I've introduced arrays in a separate exercise (together with string passing). (Actually, the whole thing is kind of rewritten, to make it less wordy and more about completing exercises.) I think the discussion in the chart exercise should also be updated to cover object unrolling, and plan on introducing native->Java performance separately (maybe as a profiler lab, so that people can try it themselves).

dmitry-timofeev commented on book/src/introduction.md:
```diff
@@ -0,0 +1,168 @@
+
+There are plenty of reasons why you should consider Rust if you need to write
+native code that works with Java. Most notably, you can separate JNI code into a
+specific layer, so that all benefits you get from Rust in normal application or
```
We may also say that unlike C, jni-rs offers some safer to use abstractions.

For example, there are Executor that ensures native threads get detached when they are destroyed; GlobalRef/AutoLocal/AutoByteArray that will always delete the ref; with_local_frame that won't let you miss "Pop", etc. Most of them could be implemented equally well in C++, some safety features, however, are unique to Rust (like not being able to share JNIEnv/JObject across threads).

For C++, there is https://github.com/mapbox/jni.hpp, though we didn't use that, because we had to integrate a _Rust_ framework with Java :)

johnwalker replied:
Thanks, I've updated the tone and discussion here, so that it talks more about jni-rs abstractions. The later exercises/solutions should also include some discussion of the drop guarantees, since they are really useful.

dmitry-timofeev commented on book/src/linking_and_dividing.md:
```diff
@@ -0,0 +1,241 @@
+3. Copy the C function name out of the header file it produced.
+
+Upon completing these steps, you will see a file called
+`com_github_jni_rs_jnibook_NativeAPI.h` that contains the following:
```
Nit: We probably shall not use com.github namespace. For a tutorial Java code that isn't uploaded on Central, we may use a shorter name, like jni_rs.jnibook or just jnibook.

johnwalker replied:
Sounds good to me. Maybe jnirsbook?

dmitry-timofeev replied:
👍 , or "jni_rs_book" so that it's slightly more readable.

dmitry-timofeev commented on book/src/linking_and_dividing.md:
````diff
@@ -0,0 +1,241 @@
+) {}
+```
+
+With that, we're done. You'll find thats exactly what the Rust starter code
````
```suggestion
With that, we're done. You'll find that’s exactly what the Rust starter code
```

dmitry-timofeev reviewed (commented):
Awesome job! I've reviewed all md files, and most other files.

johnwalker commented:
@dmitry-timofeev Thanks for your thoughtful response :) I'll work through your feedback today and tomorrow.

johnwalker commented:
Also, I'd certainly like to help out more. I'll subscribe to PRs/issues.

dmitry-timofeev commented:
Thanks a lot! 

Do you need a board to keep track of any smaller tasks on this tutorial (like this one: https://github.com/jni-rs/jni-rs/projects/1)? It might be more convenient than a checklist or full-blown issues.

johnwalker commented:
That could be extremely helpful. Would you make one? I can add some tasks in case folks want to contribute.

dmitry-timofeev commented:
Sure: https://github.com/jni-rs/jni-rs/projects/2

dmitry-timofeev commented on book/src/counter_solution.md:
````diff
@@ -0,0 +1,382 @@
+resources.
+
+```java
+// Autocloseable is used to allow try-with-resources
+class Counter implements Autocloseable {
````
```suggestion
// AutoCloseable is used to allow try-with-resources
class Counter implements AutoCloseable {
```

dmitry-timofeev replied:
And below

dmitry-timofeev commented on book/src/counter_solution.md:
I wouldn't say that both are necessarily problems:
(3) — not an issue unless the class is specified as thread-safe (i.e., it is a misuse to call increment and close from different threads).
(4) — depending on how the class is used, it might be a valid choice to make it `AutoCloseable` and document as one that must be closed by its users (e.g., using try-with-resources). In some cases, it will get unwieldly (many objects; or long-living objects), and other solutions would be better (`Cleaner`s or context-based approaches).

dmitry-timofeev commented on book/src/counter_solution.md:
```suggestion
1. `increment` and `get` may use the counter after it's been freed.
```

dmitry-timofeev commented on book/src/counter_solution.md:
```suggestion
5. it's not thread-safe; it's possible to call `get` during `destroy`, for
```

dmitry-timofeev commented on book/src/error_handling.md:
```diff
@@ -31,27 +49,116 @@ pub extern "system" fn Java_com_github_jni_1rs_jnibook_NativeAPI_divide(
+                    // don't throw another one.
+                    jni::errors::Error::JavaException => {},
+                    _ => {
+                        let _ = env.throw_new("java/lang/Exception", e.to_string());
```
It is not enough to prevent exception masking: you might get some error type that is not `jni::errors::Error::JavaException` _and_ have a pending exception in Java. The only reliable way is to check if there is an exception pending already:

https://github.com/exonum/exonum-java-binding/blob/a6627aa00d10bc8175d68392e9156fbf99f48bfe/exonum-java-binding/core/rust/src/utils/errors.rs#L162

dmitry-timofeev replied:
Possibly, even before downcast (see the comment below)

johnwalker replied:
I think I see your point - if we have some nested calls that translate a `JavaException` into some other value or type, this will fail.

dmitry-timofeev replied:
Possibly; we can expect anything from BL code :-) 

Also, just as with the module that controls the longs flying between Java and Rust to be valid pointers to objects of correct types (aka resource-manager); this method might be needed to many projects. If we can come up with a reasonably generic and performant one, we can include it in the library, as a separate PR.

dmitry-timofeev commented on book/src/error_handling.md:
```diff
@@ -31,27 +49,116 @@ pub extern "system" fn Java_com_github_jni_1rs_jnibook_NativeAPI_divide(
+            } else {
+            // This branch handles all other error types, such as those returned by
+            // all non-JNI code
+                let _ = env.throw_new("java/lang/Exception", e.to_string());
```
Since it appears as a library method, using _checked_ "java/lang/Exception" is not generally applicable (as it is valid to have native method declarations without checked exceptions). RuntimeException seems more appropriate.

dmitry-timofeev replied:
I wonder if there might be client code that returns non-jni error, but has a pending Java exception (e.g., by ignoring a jni error that they got; and getting an error from somewhere else).

johnwalker replied:
Thanks :)

Right, `RuntimeException` is much better. A colleague also pointed out that we'd need to decorate the native functions with `throws Exception`, which wasn't done in the examples. I'll update this to `RuntimeException`

I agree, this doesn't handling pending exceptions properly. Any unchecked (doesn't check for an exception, when an exception is possible) JNI method can lead to problems. I'll review exonum and see if we can get a better pattern here.

dmitry-timofeev replied:
> Any unchecked (doesn't check for an exception, when an exception is possible) JNI method can lead to problems

I think all methods that are required by the spec do check for exceptions (if not — that's a bug). I mostly though of some strange client code that, for example, throws an exceptions/invokes a method that throws an exception, and does not return that error straight away. Hopefully, you can just pull this check a little higher to cover any such cases. On top of that, if an actual error does not match JniError:Exception when there is a pending excpetion in Java, we can log such a discrepancy.

johnwalker replied:
Sounds good. I think the strange client code is likely common, so hopefully we can figure something out here. I still need to experiment with the jni-rs APIs to understand a little bit better.

One case I'm thinking about is dependent on jni-rs APIs. If call_method calls some Java method that throws an Exception, would the result be `Err` or `Ok`? Intuitively, I'd expect it to be `Ok`, and so some additional checks are necessary that the writing doesn't advertise.

johnwalker replied:
Perhaps this is an example?

```
try_java(|| {
  env.call_method(obj, "throwAnException", "()V"); // ignore exception.
  env.call_method(obj, "doSomethingElse", "()V"); // exception still pending
});
```

johnwalker replied:
Maybe something like this instead?
```rust

fn try_java<F, T>(env: JNIEnv, error_value: T, f: F) -> T
    where
        F: FnOnce() -> Result<T, anyhow::Error>,
{
    match f() {
        Ok(s) => s,
        Err(e) => {
            // Only throw an exception is one isn't already pending.
            if !env.exception_check().unwrap() {
                env.throw_new("java/lang/RuntimeException", e.to_string())
            }
            error_value
        }
    }
}
```

dmitry-timofeev replied:
> So far, I haven't found a JNI function call that returns Ok when it finishes with an exception pending

There might be some examples that may return Error, but of a different type (e.g., if they check an int error code first). But generally, methods that wrap JNI APIs that can throw exceptions use a macro that checks them for you (so that the client code rarely, if ever, have to). For example:
https://github.com/jni-rs/jni-rs/blob/master/src/wrapper/macros.rs#L21

> I think this is probably OK to start with.

+1; exception_check must not be particularly expensive: https://github.com/jni-rs/jni-rs/blob/master/benches/api_calls.rs#L206-L217

dmitry-timofeev commented on book/src/introduction.md:
```diff
@@ -2,167 +2,107 @@
 
-Assuming that the test succeeded, you're done! Consider setting up the same
-configuration for your Java editor.
+Then verify that your path is listed Exception message. If it's not, then you
```
```suggestion
Then verify that your path is listed in the Exception message. If it's not, then you
```

johnwalker commented on book/src/chart_exercise.md:
```diff
@@ -0,0 +1,102 @@
+### Reading Java Array Lengths
+[`JNIEnv::get_array_length(&self, array: jarray) ->
+Result<jsize>`](https://docs.rs/jni/0.18.0/jni/struct.JNIEnv.html#method.get_array_length)
+can be used to read the length of Java Arrays. Note that it can be better to
+pass the array length together with the array, down to the native layer, when
+that is an option, since it saves an upcall.
```
That's a good point, it's very likely that some JNI calls are less expensive than others, and that getting length is common enough that some VMs will handle it relatively well. I'll remove that suggestion, because it depends on the implementation.

dmitry-timofeev replied:
FYI, there are some benches that could be extended. Mostly they cover native -> Java method invocation using approaches on the various level of abstraction (from "pass class and method names as strings" to "look up the class ref and method ID and reuse between repeated invocations").

dmitry-timofeev replied:
https://github.com/jni-rs/jni-rs/blob/master/benches/api_calls.rs

johnwalker replied:
I'm tempted to include a reference to the benches in the materials (perhaps in future topics). What do you think?

dmitry-timofeev replied:
Agree, the set of tests already gives insights into the price of some features or coding patterns; it could be extended with more tests. It's good for users to understand the trade-offs of various options we provide, on their JVM version.

johnwalker commented:
Also, I think I don't have permissions to add cards to this board: https://github.com/jni-rs/jni-rs/projects/2

dmitry-timofeev commented on book/src/future_topics.md:
```diff
@@ -9,9 +9,10 @@ jni-rs has additional abstractions that haven't been discussed, like executors.
-code to Java, and it's possible to cache class/methodIds/fieldIds.
+There are also performance optimizations that haven't been covered. For example,
+it is more efficient to issue calls from Java to native code than from native
+code to Java, and caching class/methodIds/fieldIds can reduce the number of
```
Possibly, this section could be useful to link: https://docs.rs/jni/0.18.0/jni/struct.JNIEnv.html#checked-and-unchecked-methods

dmitry-timofeev commented:
Weird; I guess the "Triage" role is not enough; please try with "Write". I've shared the feedback to GH team.

Progdrasil commented on book/src/error_handling.md:
```diff
@@ -0,0 +1,173 @@
+            // all non-JNI code
+                let _ = env.throw_new("java/lang/RuntimeException", e.to_string());
+            }
+            default_value
```
Found this MR while looking for information on error handling with this crate.

Found this mistake where you call a value `default_value` whilst the parameter is actually `dummy_value` so either

```suggestion
            dummy_value
```

johnwalker replied:
Thanks, there are a few typos in the book. I'm going to update this tomorrow with a set of changes that includes the fix. (I switched the code samples to use Anchor tags, so that we can catch these syntactic errors more easily in the future).

Thank you for commenting that you found this though, I'd appreciate you examining the diff once it's out in case I made other mistakes.

Progdrasil commented on book/src/error_handling.md:
````diff
@@ -0,0 +1,173 @@
+your own Error hierarchies, which could have a different logic.
+
+```rust
+fn try_java<F, T>(env: JNIEnv, dummy_value: T, f: F) -> T
````
or this,
```suggestion
fn try_java<F, T>(env: JNIEnv, default_value: T, f: F) -> T
```

johnwalker replied:
This is the one it uses after the correction :)

johnwalker commented:
Alright, apologies for the big diff. I probably should have put the structural changes for getting more help from the compiler and the edits in separate commits.

1. Addressed previous reviewer comments, updated all links to `0.19.0`, and updated the jni-rs dependency.
2. Added feature flags to a completed project, the compiler will test more for us. There are still a few sections that haven't moved there yet, but they tend to be small.
3. Added `<details>` tags to the Counter discussion section, and 1 question about using Box.
4. Added `catch_unwind` to the error_handling section. Before we just said it was possible, now we actually tell how to do it.
5. Use `expect` instead of `unwrap` in `try_java` for `throw_exception`. This is useful in case someone writes the code out manually, and uses an exception class that does not exist.
6. Fix a bug in `checkAvailability`, where `Error` is wrapped with `RuntimeException`.
7. Make the `ChartRenderer` method static. The constructor was distracting.
8. Remove the Counter solution section for now. The set of changes is big enough. I'll bring this into a separate PR.
9. Replace Exception with exception with lowercase to reduce possible confusion with checked exceptions, where appropriate.
10. Suggest that the reader use cargo watch together with gradle, to rerun tests and builds. This helps users without Java IDEs.
11. Various editing improvements.

dmitry-timofeev commented on book/src/introduction.md:
```diff
@@ -8,18 +8,25 @@ intended for people that are new to JNI, and would like to learn through
+IPC/RPC). Eventually, [Project
+Panama](https://openjdk.java.net/projects/panama/) will make it [unnecessary to
+write JNI-like
+bindings](https://github.com/openjdk/panama-foreign/blob/foreign-jextract/doc/panama_jextract.md)
```
The Panama features include:
- Linking to a native library — it lands in 16 as an incubator feature and already provides an alternative to JNI. Instead of implementing native methods, users will be able to assemble some MethodHandles: 
  - https://openjdk.java.net/jeps/389
  - https://github.com/openjdk/panama-foreign/blob/foreign-jextract/doc/panama_ffi.md#appendix-full-source-code
- jextract — will assemble these MethodHandles for you based on the library header.

johnwalker replied:
Thanks :) I've updated it to include these, let me know what you think.

dmitry-timofeev commented on book/src/counter_discussion.md:
It could be unclear to the reader what to use :) 

The corresponding code in EJB returns a reference to avoid this situation: https://github.com/exonum/exonum-java-binding/blob/a6627aa00d10bc8175d68392e9156fbf99f48bfe/exonum-java-binding/core/rust/src/handle/mod.rs#L39-L55 

Possibly, it might make sense to first add this library code; and then just use it in the solution.

johnwalker replied:
I agree, it would be much better if it offered a solution here. I've backed off from providing one because I haven't found a reference for the invariants the programmer must ensure hold when they perform the cast. The unsafe code guidelines/rustonomicon/rust book didn't have anything that clicked here.

There could be one other alternative, which is dereferencing a raw pointer instead.

johnwalker commented on book/projects/starter/jnibookrs/src/division.rs:
```diff
@@ -0,0 +1,72 @@
+// ANCHOR: imports
```
Ah, I'm crazy and didn't notice. Fixing

johnwalker replied:
Fixed

johnwalker replied:
(It was also just that one with this property)

johnwalker commented on book/src/future_topics.md:
Fixed, thanks

rib commented:
Hi @johnwalker 

Sorry that it's literally been years since you first started working on this and then things went quiet before there was a final resolution for this work.

Especially in the last year or so the project hasn't had lots of active development and I've only quite recently taken over maintaining the project (though still rather ad-hoc, not with a lot of dedicated time) and have been working through the backlog of issues and PRs to figure out how to address them.

I think @dmitry-timofeev was most familiar with the contents of your work and was doing a good job at providing feedback but it looks like that review process lost steam around February last year.

It looks like there are some good gems of information in the book even if it wasn't entirely complete and it would be good if those things could be shared somehow.

Unfortunately it seems like adding a book to this repo probably just ends up raising the bar for publishing something because all the content needs to have buy in + review from the project maintainers and we probably need to be a more considerate of how we would plan to maintain it over time (e.g. considering evolving API changes).

On the other hand if the same content were published independently in some form you could still seek review/feedback but there would be no other constraints on how / when you publish the content. E.g. Maybe some of the content could be adapted into blog posts.

We could also sign post from this project e.g. to blog posts or a personal repo hosting this book in a way that still means people could discover it via this project.

I have to admit that I don't think I'm going to be in a position to pick up an active review process for this much content and so I don't currently see that there will be much progress here unless another maintainer comes along with more time to dedicate.

If you have a moment, it'd be good to hear your thoughts.

rib commented:
Okey, I'm going to close this issue for now. Sorry this work didn't land in the end. I'm still subscribed to this issue so should see if there are any follow up comments while it's closed.

johnwalker commented:
Thanks for following up, it likely needs a bit of a refresh. It sounds like you're OK with this being published separately, perhaps with some refresh. And maybe in the future we can re-evaluate whether to transition ownership. What do you think?
